components/uniset/src/uniset.rs

```diff
@@ -0,0 +1,315 @@
+            .sum();
+        s as usize
+    }
+
+    /// Returns whether or not the UnicodeSet is empty
+    pub fn is_empty(&self) -> bool {
+        self.inv_list.is_empty()
+    }
+
+    /// Wrapper for contains
+    ///
+    /// Returns an Option as to whether or not it is possible for the query to be contained
+    fn contains_impl(&self, query: u32) -> Option<usize> {
+        match self.inv_list.binary_search(&query) {
+            Ok(pos) => {
+                if pos % 2 == 0 {
+                    Some(pos)
+                } else {
+                    None
+                }
+            }
+            Err(pos) => {
+                if pos % 2 != 0 && pos < self.inv_list.len() {
+                    Some(pos)
+                } else {
+                    None
+                }
+            }
+        }
+    }
+
+    /// Checks to see the query is in the UnicodeSet
+    ///
+    /// Runs a binary search in `O(log(n))` where `n` is the number of start and end points
+    /// in the set using `std` implementation
+    ///
+    /// Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSet;
+    /// use std::convert::TryFrom;
+    /// let example_list = vec![65, 67, 68, 69];
+    /// let example = UnicodeSet::try_from(example_list).unwrap();
+    /// assert!(example.contains('A'));
+    /// assert!(!example.contains('C'));
+    /// ```
+    pub fn contains(&self, query: char) -> bool {
+        self.contains_impl(query as u32).is_some()
+    }
+
+    /// Checks to see if the range is in the UnicodeSet, returns a Result
+    ///
+    /// Runs a binary search in `O(log(n))` where `n` is the number of start and end points
+    /// in the set using `std::vec::Vec` implementation Only runs the search once on the `start`
+    /// parameter, while the `end` parameter is checked in a single `O(1)` step
+    ///
+    /// Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSet;
+    /// use std::convert::TryFrom;
+    /// let example_list = vec![65, 67, 68, 69];
+    /// let example = UnicodeSet::try_from(example_list).unwrap();
+    /// assert!(example.contains_range(&('A'..'C')));
+    /// assert!(example.contains_range(&('A'..='B')));
+    /// assert!(!example.contains_range(&('A'..='C')));
+    /// ```
+    ///
+    /// Surrogate points (`0xD800 -> 0xDFFF`) will return false if the Range contains them but the
+    /// UnicodeSet does not.
+    ///
+    /// Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSet;
+    /// use std::{convert::TryFrom, char::from_u32};
+    /// let check = from_u32(0xD7FE).unwrap() .. from_u32(0xE001).unwrap();
+    /// let example_list = vec![0xD7FE, 0xD7FF, 0xE000, 0xE001];
+    /// let example = UnicodeSet::try_from(example_list).unwrap();
+    /// assert!(!example.contains_range(&(check)));
+    /// ```
+    pub fn contains_range(&self, range: &impl RangeBounds<char>) -> bool {
+        let (from, till) = deconstruct_range(range);
+        if from >= till {
```

Comment:
is that possible for `deconstruct_range` to return `from` larger/equal to `till`?

Reply:
an `assert!` could work off the top of my head ~ I think that I want to keep the `util` functions definite (avoiding `Option` and `Result`) but let me know if just doing one of those might be best over an `assert!` call

Reply:
`assert!` should be used if you can't limit the input into a public function but certain input is unacceptable (you want to panic if someone passed `divide(2, 0)`).
`Option` should be used if the output may be `None` and it's a legal value. `Result` should be used if the output may be `Error` and it's okay (for example, parsing).

In this case, I believe the *output* should be bool, because you don't need to return any additional information.
On the *input* you're taking a range, and then calling an internal function on that range.

What I'm not sure if it is possible for that internal function to return `from >= till` is a possible value to be returned out of `deconstruct_range`.

If it is, but should never happen, you can use `debug_assert!` which is a good tradeoff between "I had not found a better way to ensure that" and "It should never happen".

Reply:
> is that possible for `deconstruct_range` to return `from` larger/equal to `till`?

Yes.  See previous discussion: https://github.com/unicode-org/icu4x/pull/122#discussion_r446445202

Reply:
sounds good.

Reply:
I think the highlighted code may be confusing people a little:  

`deconstruct_range` returns a tuple `(from, till)`, but we want to keep this infallible while maintaining the check from `if from >= till`, but we are trying to return the tuple still. At first the idea of returning an empty range from `deconstruct_range` was entertained, but that was dropped since the check for the empty range will still need to happen...  I'm also against doing `debug_assert!` because it will panic rather than just saying that the range isn't possible...

If keeping the `if from >= till` is detrimental the architecture of `deconstruct_range` and in turn the `try_from` statements could be changed, but I think it just offers slight compartmentalization that isn't of great concern